calendarsubscription/store.go

```diff
@@ -0,0 +1,189 @@
+	p := &util.Prepare{DB: db, Ctx: ctx}
+
+	return &Store{
+		db:      db,
+		findOne: p.P(`SELECT * FROM user_calendar_subscriptions cs WHERE cs.id = $1`),
+		create: p.P(`
+				INSERT INTO user_calendar_subscriptions (user_id, id, name, config, schedule_id, disabled)
+				VALUES ($1, $2, $3, $4, $5, $6)
+			`),
+		update:     p.P(`UPDATE user_calendar_subscriptions SET name = $3, disabled = $4, config = $5 WHERE id = $1 AND user_id = $2`),
+		delete:     p.P(`DELETE FROM user_calendar_subscriptions WHERE id = any($1) AND user_id = $2`),
+		findAll:    p.P(`SELECT * FROM user_calendar_subscriptions WHERE user_id = $1`),
+		findOneUpd: p.P(`SELECT id, name, user_id, disabled, config FROM user_calendar_subscriptions WHERE id = $1 FOR UPDATE`),
+	}, p.Err
+}
+
+func wrapTx(ctx context.Context, tx *sql.Tx, stmt *sql.Stmt) *sql.Stmt {
+	if tx == nil {
+		return stmt
+	}
+	return tx.StmtContext(ctx, stmt)
+}
+
+// FindOne will return a single calendar subscription for the given id.
+func (b *Store) FindOne(ctx context.Context, id string) (*CalendarSubscription, error) {
+	err := permission.LimitCheckAny(ctx, permission.All)
+	if err != nil {
+		return nil, err
+	}
+
+	err = validate.UUID("CalendarSubscriptionID", id)
+	if err != nil {
+		return nil, err
+	}
+
+	var cs CalendarSubscription
+	err = b.findOne.QueryRowContext(ctx, id).Scan(&cs.ID, &cs.Name, &cs.UserID, &cs.LastAccess, &cs.Disabled, &cs.ScheduleID, &cs.Config)
+	if err != nil {
+		return nil, err
+	}
+	return &cs, nil
+}
+
+// CreateSubscriptionTx will return a created calendar subscription with the given input.
+func (b *Store) CreateSubscriptionTx(ctx context.Context, tx *sql.Tx, cs *CalendarSubscription) (*CalendarSubscription, error) {
+	cs.UserID = permission.UserID(ctx)
+	err := permission.LimitCheckAny(ctx, permission.Admin, permission.MatchUser(cs.UserID))
+	if err != nil {
+		return nil, err
+	}
+	cs.ID = uuid.NewV4().String()
+	cs, err = cs.Normalize()
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = wrapTx(ctx, tx, b.create).ExecContext(ctx, cs.UserID, cs.ID, cs.Name, cs.Config, cs.ScheduleID, cs.Disabled)
+	if err != nil {
+		return nil, err
+	}
+	return cs, nil
+}
+func (b *Store) FindOneForUpdateTx(ctx context.Context, tx *sql.Tx, id string) (*CalendarSubscription, error) {
+	var userID = permission.UserID(ctx)
+	err := permission.LimitCheckAny(ctx, permission.Admin, permission.MatchUser(userID))
+	if err != nil {
+		return nil, err
+	}
+	err = validate.UUID("CalendarSubscriptionID", id)
+	if err != nil {
+		return nil, err
+	}
+
+	var cs CalendarSubscription
+
+	row := wrapTx(ctx, tx, b.findOneUpd).QueryRowContext(ctx, id)
+	err = row.Scan(&cs.ID, &cs.Name, &cs.UserID, &cs.Disabled, &cs.Config)
+	if err != nil {
+		return nil, err
+	}
+	return &cs, nil
+}
+
+// UpdateSubscriptionTx updates a calendar subscription with given information.
+func (b *Store) UpdateSubscriptionTx(ctx context.Context, tx *sql.Tx, cs *CalendarSubscription) error {
```

Comment:
This can be just `UpdateTx` since we're already `calendarsubscription.Store` and only have one type